plugins/scope_navigate.py

```diff
@@ -0,0 +1,51 @@
+import sublime
+import sublime_plugin
+
+# Related Reading:
+#     https://forum.sublimetext.com/t/keybinding-to-find-a-particular-set-of-characters/49064
+#
+# This command allows you to easily jump the cursor to the next or previous
+# location of a specific "scope" in the file. This is a bit more advanced as a
+# navigation method, but would allow you to for example create a key binding
+# that always jumps between comments in files, even when switching languages.
+#
+# Use 'Tools > Developer > Show Scope Name' to determine the scope to use here.
+
+
+class ScopeNavigateCommand(sublime_plugin.TextCommand):
+    """
+    Jump the selection in the file to the next or previous location of the
+    given scope based on the current cursor location. The search direction is
+    controlled by the forward argument, and will wrap around the ends of the
+    buffer.
+    """
+    def run(self, edit, scope, forward=True):
+        # Find the locations where this scope occurs; leave if none
+        regions = self.view.find_by_selector(scope)
+        if not regions:
+            return
+
+        # Get a starting point for our search, and where we should jump to if
+        # there are no matches in the specified direction.
+        point = self.view.sel()[0].b
+        fallback = regions[-1] if not forward else regions[0]
+
+        # Remove all selections.
+        self.view.sel().clear()
+
+        # Look in the given direction for the first match from the current
+        # position; if one is found jump there.
+        pick = lambda p: (point < p.a) if forward else (point > p.a)
+        for pos in regions if forward else reversed(regions):
```

Comment:
nice code re-use here :)

btw this scrap reminds me of one of mine:

```py
import sublime
import sublime_plugin


class SelectByScopeSelectorCommand(sublime_plugin.TextCommand):
    def run(self, edit, scope_selector=None, last_scope_only=True, auto_select=False):
        if not scope_selector:
            scope_at_first_caret = self.view.scope_name(self.view.sel()[0].a)
            if last_scope_only:
                scope_at_first_caret = scope_at_first_caret.split()[-1]

            if auto_select:
                scope_selector = scope_at_first_caret
            else:
                self.view.window().show_input_panel('Scope Selector', scope_at_first_caret, lambda value: self.view.run_command('select_by_scope_selector', { 'scope_selector': value }), None, None)
                return

        regions = self.view.find_by_selector(scope_selector)
        if regions:
            self.view.sel().clear()
            self.view.sel().add_all(regions)
            self.view.show_at_center(self.view.sel()[0])
        else:
            self.view.window().status_message('Unable to find anything matching selector "' + scope_selector + '"')
```

one thing I like about mine is - you can default the scope selector to the scope under the caret. And prompt the user to enter it. And that I refer to "scope selectors" instead of scopes ;)
Anyway, just food for thought - I like what you've done here and how it works in a similar way to another scrap.

Reply:
That's interesting! Is that in the scraps repo? A quick search didn't find it, but things are going a little weird this evening in general...

Reply:
Nope, private atm / just a stackoverflow answer I believe. I also need to get round to posting stuff on this repo!